django/contrib/auth/middleware.py

```diff
@@ -34,6 +39,68 @@ def process_request(self, request):
         request.auser = partial(auser, request)
 
 
+class LoginRequiredMiddleware:
+    """
+    Middleware that forces all views to require login by default.
+
+    Views that have login_not_required decorator or LoginNotRequiredMixin mixin
+    will be able to pass through without this validation. Otherwise, it will
+    redirect user to login or admin's login if accessing admin site.
+    """
+
+    sync_capable = True
+    async_capable = True
+
+    def __init__(self, get_response):
+        if get_response is None:
+            raise ValueError("get_response must be provided.")
+        self.get_response = get_response
+        self.async_mode = iscoroutinefunction(self.get_response)
+        if self.async_mode:
+            markcoroutinefunction(self)
+        super().__init__()
+
+    def __call__(self, request):
+        if not hasattr(request, "user"):
+            raise ImproperlyConfigured(
+                "The Django login required middleware requires authentication "
+                "middleware to be installed. Edit your MIDDLEWARE setting to "
+                "insert "
+                "'django.contrib.auth.middleware.AuthenticationMiddleware' "
+                "before "
+                "'django.contrib.auth.middleware.LoginRequiredMiddleware'."
+            )
+        if self.async_mode:
+            return self.__acall__(request)
+        self.user_is_authenticated = request.user.is_authenticated
+        return self.get_response(request)
+
+    async def __acall__(self, request):
+        self.user_is_authenticated = (await request.auser()).is_authenticated
+        return await self.get_response(request)
+
+    def process_view(self, request, view_func, view_args, view_kwargs):
+        if self.user_is_authenticated:
+            return None
+
+        view_class = getattr(view_func, "view_class", None)
+        if view_class and not getattr(view_class, "login_required", True):
+            return None
+
+        if not getattr(view_func, "login_required", True):
+            return None
+
+        try:
+            if request.path.startswith(reverse("admin:index")):
+                return redirect_to_login(
+                    request.get_full_path(), reverse("admin:login")
+                )
+        except NoReverseMatch:
+            pass
+
+        return redirect_to_login(request.get_full_path())
```

Comment:
Defining `process_view` will still result in context switches :(

Ref this code:

https://github.com/django/django/blob/main/django/core/handlers/base.py#L77-L81

I think you need to rename this function to something else (so it doesn't get caught by that auto-rewrite code) and then call it from `__call__` / `__acall__`

Reply:
Yes, I thought nothing in `process_view` was io bound so it didn't really matter much if it got wrapped in `sync_to_async`? 

The way I see to achieve no context switch here is to have our local [resolve_request](https://github.com/django/django/blob/32d70b2f55b1f74736fd11bc8efce890ad5fa2f0/django/core/handlers/base.py#L300). But I don't know if its worth it? What do you think?

Reply:
> Yes, I thought nothing in process_view was io bound so it didn't really matter much if it got wrapped in sync_to_async?

Wrapping in sync_to_async still causes a context switch / performance hit :(

> Yes, I thought nothing in process_view was io bound so it didn't really matter much if it got wrapped in sync_to_async?

The way I see to achieve no context switch here is to have our local [resolve_request](https://github.com/django/django/blob/32d70b2f55b1f74736fd11bc8efce890ad5fa2f0/django/core/handlers/base.py#L300). But I don't know if its worth it? What do you think?

I don't think that's necessary, I think you can just change the name to *not* hit any of the names that will be rewritten and then call this function from `__call__` / `__acall__`.

Reply:
Yes, but the thing is `__call__`/`__acall__` is not provided a `view_func`/`callback` it is only passed to `process_view` [here](https://github.com/django/django/blob/32d70b2f55b1f74736fd11bc8efce890ad5fa2f0/django/core/handlers/base.py#L186) only after [resolving the request](https://github.com/django/django/blob/32d70b2f55b1f74736fd11bc8efce890ad5fa2f0/django/core/handlers/base.py#L181).

So the way forward I see is:
1) Using a private resolve_request (this gets called twice here and in handler)
2) Changing how handler works to support native async (not using same `process_view` name for both sync and async)
3) Finding a way to not use `view_func`/`callback` at all (I didn't look much into it yet)

Am I missing something?

Reply:
> Changing how handler works to support native async (not using same process_view name for both sync and async)

I'm kinda in favor of this idea, but I can understand your hesitancy towards doing it. I think for now we can leave this as you have implemented in this PR and I'll think about how we can solve it better in the general case.

Reply:
+1 for changing the handler, but in a separate ticket and PR.